File: 01. Arrays/08. Majority Element I/main.cpp
#include <bits/stdc++.h>
using namespace std;

// Approach 1: Brute Force
int majorityElementBruteForce(vector<int>& nums) {
    int n = nums.size();
    for (int i = 0; i < n; i++) {
        int count = 0;
        for (int j = 0; j < n; j++) {
            if (nums[j] == nums[i]) count++;
        }
        if (count > n / 2) return nums[i];
    }
    return -1;
}

// Approach 2: Sorting
int majorityElementSorting(vector<int>& nums) {
    sort(nums.begin(), nums.end());
    return nums[nums.size() / 2];
}

// Approach 3: HashMap
int majorityElementHashMap(vector<int>& nums) {
    unordered_map<int, int> freq;
    for (int num : nums) freq[num]++;
    for (auto [key, value] : freq) {
        if (value > nums.size() / 2) return key;
    }
    return -1;
}

// Approach 4: Boyer-Moore Voting Algorithm (Optimal)
int majorityElementBoyerMoore(vector<int>& nums) {
    int candidate = 0, count = 0;
    for (int num : nums) {
        if (count == 0) candidate = num;
        count += (num == candidate) ? 1 : -1;
    }
    return candidate;
}

// Main Function
int main() {
    vector<int> nums = {3, 2, 3};
    cout << "Brute Force: " << majorityElementBruteForce(nums) << endl;
    cout << "Sorting: " << majorityElementSorting(nums) << endl;
    cout << "HashMap: " << majorityElementHashMap(nums) << endl;
    cout << "Boyer-Moore: " << majorityElementBoyerMoore(nums) << endl;
    return 0;
}

File: 01. Arrays/08. Majority Element I/main.java
import java.util.Arrays;
import java.util.HashMap;

class main {
    // Brute Force
    public static int majorityElementBruteForce(int[] nums) {
        int n = nums.length;
        for (int i = 0; i < n; i++) {
            int count = 0;
            for (int j = 0; j < n; j++) {
                if (nums[j] == nums[i]) count++;
            }
            if (count > n / 2) return nums[i];
        }
        return -1;
    }

    // Sorting
    public static int majorityElementSorting(int[] nums) {
        Arrays.sort(nums);
        return nums[nums.length / 2];
    }

    // HashMap
    public static int majorityElementHashMap(int[] nums) {
        HashMap<Integer, Integer> freq = new HashMap<>();
        for (int num : nums) freq.put(num, freq.getOrDefault(num, 0) + 1);
        for (int key : freq.keySet()) {
            if (freq.get(key) > nums.length / 2) return key;
        }
        return -1;
    }

    // Boyer-Moore Voting Algorithm (Optimal)
    public static int majorityElementBoyerMoore(int[] nums) {
        int candidate = 0, count = 0;
        for (int num : nums) {
            if (count == 0) candidate = num;
            count += (num == candidate) ? 1 : -1;
        }
        return candidate;
    }

    // Main Function
    public static void main(String[] args) {
        int[] nums = {3, 2, 3};
        System.out.println("Brute Force: " + majorityElementBruteForce(nums));
        System.out.println("Sorting: " + majorityElementSorting(nums));
        System.out.println("HashMap: " + majorityElementHashMap(nums));
        System.out.println("Boyer-Moore: " + majorityElementBoyerMoore(nums));
    }
}
